src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,151 @@
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        if (prev == null) {
+            prev = next = new Node<>(null, value, null);
+        } else if (index == 0) {
+            prev.prev = new Node<>(null, value, prev);
+            prev = prev.prev;
+        } else if (index == size) {
+            next.next = new Node<>(next, value, null);
+            next = next.next;
+        } else {
+            Node<T> current = prev;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+            current.prev.next = new Node<>(current.prev, value, current);
+            current.prev = current.prev.next;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexOutOfBoundsException(index);
+        Node<T> current = prev;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexOutOfBoundsException(index);
+        Node<T> current = prev;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T oldValue = current.item;
+        current.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexOutOfBoundsException(index);
+        if (prev == null) {
+            throw new RuntimeException("List is empty");
+        }
+        Node<T> current;
+        if (index == 0) {
+            current = prev;
+        } else if (index == size - 1) {
+            current = next;
+        } else {
+            current = prev;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        }
+        unlink(current);
+        return current.item;
     }
 
     @Override
-    public boolean remove(T object) {
+    public boolean remove(T node) {
+        Node<T> current = prev;
+        while (current != null) {
+            if (areEqual(node, current.item)) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean areEqual(T a, T b) {
+        return (a == null && b == null) || (a != null && a.equals(b));
+    }
```

Comment:
for this exists Objects.equals(a, b)

Reply:
we shouldn't use class Object.

Reply:
I also already met with that. agree with you